### PR TITLE
Emoji fix

### PR DESCRIPTION
**Pull request by juancampa**
## Problems
There were multiple problems with emoji

 1. Opening the emoji dialog was possible but it appeared at a weird location. This was due to the textarea we use for input was not being kept in sync with the cursor position unless a composition (i.e. IME input) was underway.
 1. No emoji was received by the terminal.

## Solutions

 1. Now the textarea always moves with the cursor.
 1. Added handler for the `input` event, only for emoji handling. Also, the textarea needs to have a `lineheight > 0` for this event to fire, at least in Chromium.

## Testing

 - [ ] MacOS
   - [x] Emojis are inserted from the emoji input popup
   - [x] The popup shows at the right location
   - [x] Pasting still works fine. Including pasting emoji
   - [x] Chrome
   - [ ] Firefox
   - [x] Safari
   - [x] IMEs still work (jp/ko/ch) including inserting emoji from Pinyin IME
 - [ ] Windows
   - [ ] Emojis are inserted from the emoji input popup
   - [ ] The popup shows at the right location
   - [ ] Pasting still works fine. Including pasting emoji
   - [ ] Chrome
   - [ ] Firefox
   - [ ] Safari
   - [x] ~~IE (support dropped)~~
   - [ ] IMEs still work (jp/ko/ch) including inserting emoji from Pinyin IME
 - [x] Electron (via Hyper)

### CODE REVIEW

**Comment by Tyriar** [comment]
The position is still broken on Windows:

win+.

image

Actual input doesn't seem to work either, but that's probably related to shell/pty problems.

image

File: yarn.lock [hunk 1]
```diff
@@ -26,6 +26,11 @@
   resolved "https://registry.yarnpkg.com/@types/chai/-/chai-3.5.2.tgz#c11cd2817d3a401b7ba0f5a420f35c56139b1c1e"
   integrity sha1-wRzSgX06QBt7oPWkIPNcVhObHB4=
 
+"@types/dom-inputevent@^1.0.3":
```
Comment (Tyriar) [inline yarn.lock]: This should be undone since we're not using anymore

File: src/common/Types.ts [hunk 2]
```diff
@@ -21,6 +21,15 @@ export interface IKeyboardEvent {
   type: string;
 }
 
+/**
+ * There's a @types/dom-inputevent but at the moment it's outdated and missing
+ * some properties (e.g. inputType). So in the meantime we're using this instead
+ */
+export interface IInputEvent {
+  readonly data?: string;
+  readonly inputType?: string;
```
Comment (Tyriar) [inline src/common/Types.ts]: It's missing from the type probably because support is bad (no Firefox), do you know why it appears to work on Firefox despite this? https://developer.mozilla.org/en-US/docs/Web/API/InputEvent/inputType#Browser_compatibility

File: src/Emoji.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
+// This function returns whether the first code point of the provided string is
+// within a range that contains emoji. Some of the codepoints in these ranges
+// are unassigned so this test is an approximation to the real thing.
+// Ranges taken from:
+//   https://stackoverflow.com/questions/30757193/
+// For a complete list of all assigned codepoints, go to:
+//   https://unicode.org/Public/emoji/11.0/emoji-data.txt
+export function isEmoji(str: string): boolean {
```
Comment (Tyriar) [inline src/Emoji.ts]: Let's move this into CharWidth.ts, this will later be renamed to Unicode*.ts so i think it makes sense to live there.

**Comment by juancampa** [comment]
> The position is still broken on Windows:

Unfortunately I don't have a good way to test on Windows at the moment. Trying to get access to a remote desktop machine.

> macOS/firefox

Looking into it, it seems to work sometimes though :thinking:

File: src/CompositionHelper.ts [hunk 4]
```diff
@@ -190,10 +190,6 @@ export class CompositionHelper {
    *   necessary as the IME events across browsers are not consistently triggered.
    */
   public updateCompositionElements(dontRecurse?: boolean): void {
-    if (!this._isComposing) {
-      return;
-    }
-
     if (this._terminal.buffer.isCursorInViewport) {
```
Comment (Tyriar) [inline src/CompositionHelper.ts]: Why did you need to remove this? I don't think it's meant to move when composing (after it becomes visible)

File: src/Emoji.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  if (str.codePointAt === undefined) {
```
Comment (Tyriar) [inline src/Emoji.ts]: Please add:

```
// TODO: Remove when IE11 support is dropped
```

File: src/common/Types.ts [hunk 6]
```diff
@@ -21,6 +21,15 @@ export interface IKeyboardEvent {
   type: string;
 }
 
+/**
+ * There's a @types/dom-inputevent but at the moment it's outdated and missing
+ * some properties (e.g. inputType). So in the meantime we're using this instead
+ */
+export interface IInputEvent {
```
Comment (Tyriar) [inline src/common/Types.ts]: Let's call this `IDomInputEvent` to be clear as `IKeyboardEvent` above is an xterm.js thing

File: src/common/Types.ts (hunk 2 above)
Comment (epicfaace) [reply src/common/Types.ts]: @Tyriar The `data` attribute is not supported on Firefox. As per my previous comment, inputting emojis somewhat works with Firefox even _without_ the code on this branch: https://github.com/xtermjs/xterm.js/issues/469#issuecomment-451789158

Comment (juancampa) [reply src/common/Types.ts]: Interesting. I'll run it with the debugger to understand what's happening on Firefox

File: src/CompositionHelper.ts (hunk 4 above)
Comment (juancampa) [reply src/CompositionHelper.ts]: Not sure I follow. If that were the case, shouldn't the `if` be `if (this._isComposing)`. My understanding is that this `if` avoids unnecessarily moving the `textarea` when is not needed (not composing anything), but I might be wrong :stuck_out_tongue_closed_eyes: 

The idea here is that we need a `textarea` to follow the cursor at all times so that the emoji popup shows at the right position (right under the text area).

Comment (juancampa) [reply src/CompositionHelper.ts]: Thinking about this a bit more, it might be better to move the positioning of the `textarea` to `Terminal.ts` since it's not a composition thing anymore

File: src/common/Types.ts (hunk 2 above)
Comment (juancampa) [reply src/common/Types.ts]: I did a bit of investigation on Firefox:

 - An alternative to `data` is to use `event.target.value` but only if we're clearing the `textarea` on every change
 - Firefox positioning of the Emoji popup is buggy in that, if you move the `textarea` around, the emoji popup's position (later, when it shows) will use the previous position of the textarea. A work-around I found is to manually set `textarea.value` to something (doesn't matter what), which seems to force firefox to recalculate the position used for emoji popup. Seems like a cache invalidation bug.
 - Firefox treats emoji input as composition while chrome doesn't :man_shrugging:

Comment (juancampa) [reply src/common/Types.ts]: - Oh, and finally, and this might be a deal breaker, there's no `inputType` on Firefox, which we're using to detect when text is inserted vs. when text is pasted. There might be a way to work around this, by treating it as a composition but I haven't looked deeper into it

Comment (epicfaace) [reply src/common/Types.ts]: If we can't easily do it reliably on Firefox, I think at least doing it on Chrome would definitely be an improvement.

**Comment by Tyriar** [comment]
@jerch do you know why if you press space twice after inserting an emoji a `.` appears?

screen shot 2019-01-26 at 12 51 59 pm

This even happens on master when you paste an emoji

File: src/CompositionHelper.ts (hunk 4 above)
Comment (Tyriar) [reply src/CompositionHelper.ts]: This function is only meant to do anything after `compositionstart` and before `compositionend`, I think you're right that maybe this should be moved out.

**Comment by Tyriar** [comment]
Ping @juancampa, you'll have difficulty keeping this page up to date in Hyper if it doesn't get merged in

**Comment by Tyriar** [comment]
Closing off since this needed a bunch of changes and has diverged a lot. A new attempt at fixing should probably start on a fresh branch.
